### DISCUSSION TITLE
Persisting `-define` flags

### DISCUSSION THREAD

**Discussion by mertalev** [ImageMagick/ImageMagick#6571] [Help] [answered]
### ImageMagick version

7.1.1.-15

### Operating system, version and so on

Debian Bookworm

### Description

I'd like to be able to set `-define dng:use-camera-wb=true` for Imagemagick as RAW images can otherwise have shifted colors ([related](https://stackoverflow.com/questions/71983812/wrong-whitebalace-colorgrade-for-arw-files-in-imagemagick-updated)). 

But since I'm not using Imagemagick directly, instead through sharp (and by extension libvips), I don't have much control over the exact command passed to Imagemagick. Short of patching the relevant code to assign true for this flag, there doesn't seem to be anything I can do.

It would be very convenient if there were some other way of configuring settings like this, whether as a config file or as a command I can run beforehand to set this as default behavior.

**Comment by snibgo** [comment]
In your StackOverflow question:

> When I open the files in ACDSee or XNView, they look like the jpg-version off the camera, but when I open them in imagemagick Display, they are much darker and more reddish.

Comparing the two images, there is also geometric distortion between the two, probably to correct for barrel distortion in the lens. Perhaps ACDSee and XNView are showing you the JPEGs that are embedded in the ARW files. If so, they will be *identical* to the JPEGs made by the camera, and have been corrected and prettified by the camera's algorithms.

By contrast, IM processes the raw data from the ARW file. By default, IM doesn't prettify images, or correct barrel distortion, etc. You can add options to do this.

> But since I'm not using Imagemagick directly, instead through sharp (and by extension libvips), I don't have much control over the exact command passed to Imagemagick. 

I don't know sharp or libvips, but if they don't give you the control you want over ImageMagick, the obvious remedy is to request that control from those products. Adding a config file or similar to IM would be a kludge.

**Reply by mertalev** [reply]
Hi, thanks for the quick response!

That post isn't mine, but I see the same color shift described by them for both ARW and CR3 (possibly others as well). This is the difference in output between disabling and enabling the `use-camera-wb` flag for an ARW file:

Disabled (default):

![_DSC2642_disabled_1440](https://github.com/ImageMagick/ImageMagick/assets/101130780/9aae2c1a-e904-4a23-bbcd-92d6d35b24a3)

Enabled:

![_DSC2642_enabled_1440](https://github.com/ImageMagick/ImageMagick/assets/101130780/bc056bc6-d79a-4636-95de-e285e2e20b9d)

A few other examples:

![Comparison1](https://user-images.githubusercontent.com/18033909/261046637-fc0b018b-ffff-4922-96e3-d0624c98e4c4.png)
![Comparison2](https://user-images.githubusercontent.com/18033909/261046808-4c8a7a17-b47b-4eab-ac63-3de80803dbbf.png)

**Reply by mertalev** [reply]
> I don't know sharp or libvips, but if they don't give you the control you want over ImageMagick, the obvious remedy is to request that control from those products. Adding a config file or similar to IM would be a kludge.

I thought of posting this on libvips, but it would require a larger API change to support this and sharp would likewise have to change its API. Since libvips only uses ImageMagick as a fallback, it's also a bit clunky to pass flags for it like that. 

It seems simpler to set this closer to where it will actually be used, especially given that ImageMagick already allows for customizing many options with .xml files.

Libraw does have a few env variables for white balance, including `LIBRAW_PROCESSING_CAMERAWB_FALLBACK_TO_DAYLIGHT` and `LIBRAW_PROCESSING_PROVIDE_NONSTANDARD_WB`, but I can't seem to find one for `use_camera_wb`.

**Comment by fmw42** [comment]
I am no expert on this, but could you not add the define information to the delegates.xml file for that type file?

**Reply by mertalev** [reply]
Maybe! I'm not sure how the delegates.xml file works. Looking at the [relevant line](https://github.com/ImageMagick/ImageMagick/blob/5bf8645839cda5279063710409c435f0b5b757f6/config/delegates.xml.in#L69), it seems to hardcode some settings I know I can configure, like file type and bit depth. Is this config file still used when using the C API with libMagick?

**Comment by snibgo** [comment]
>  Is this config file still used when using the C API with libMagick?

Yes, it is. `config.xml` is used for external delegates, which are binary programs that IM executes, such as dcraw. These days, IM uses an internal delegate libraw to process raw files. However, you can create a line in delegates.xml that calls dcraw or other program, and give it a special prefix, such as `MYRAW`.

**Reply by mertalev** [reply]
Hmm, using dcraw instead of libraw isn't ideal. 

It's a bit hacky, but patching `dng.c` like so makes the change I'd like without changing anything else:

```
@@ -339,6 +339,8 @@
     option=GetImageOption(image_info,"dng:use_camera_wb");
   if (option != (const char *) NULL)
     raw_info->params.use_camera_wb=IsStringTrue(option);
+  else
+    raw_info->params.use_camera_wb=MagickTrue;
   option=GetImageOption(image_info,"dng:use-auto-wb");
   if (option == (const char *) NULL)
     option=GetImageOption(image_info,"dng:use_auto_wb");
```

Is it possible to accomplish this in `delegates.xml`?

**Comment by snibgo** [comment, marked as answer]
Content in `delegates.xml` has no impact on what is passed to libraw.

I suspect adding an entry to `delegates.xml` wouldn't solve your problem because you would need to use the special prefix when opening the raw file, such as "`MYRAW:ABC_1234.arw`".

You can patch your version of IM, of course. But changing IM's default behaviour for all users is usually a bad idea.

**Reply by mertalev** [reply]
I agree that changing the default would be bad and confusing for people. I was wondering if configuring defaults externally could be a good way to resolve this, but that doesn't currently seem possible. It looks like I'll have to stick with the patch.

**Comment by fmw42** [comment]
you could edit a copy of the code and then compile from your source

**Reply by mertalev** [reply]
That seems to be the solution. https://github.com/immich-app/immich/pull/3806
